### ISSUE TITLE
flatpak / bubblewrap fails to start - Can't mount proc on /newroot/proc - /proc/xen mount issue

### ISSUE THREAD

**Issue by adrelanos** [QubesOS/qubes-issues#2540]
#### Qubes OS version (e.g., `R3.2`):

R3.2 with Qubes testing repository, probably Qubes R4 and above as well

#### Affected TemplateVMs (e.g., `fedora-23`, if applicable):

debian, whonix-ws, probably any Linux based template

---

### Expected behavior:

flatpak / bubblewrap working.

### Actual behavior:

flatpak / bubblewrap fails to start.

```
2016/12/25 05:35:20 firefox: Can't mount proc on /newroot/proc
2016/12/25 05:35:20 firefox: : Operation not permitted
2016/12/25 05:35:20 fatal error in the user interface: waitid: no child processes
```

### Steps to reproduce the behavior:

Update: bubblewrap Sandboxed Tor Browser no longer exists. Therefore my original bug report no longer applies. However...

@alex-mazzariol https://github.com/QubesOS/qubes-issues/issues/2540#issuecomment-278919971
> This issue affects a lot of flatpak-distributed software (i.e. MonoDevelop), not only sandboxed TOR browser. It seems that flatpak internally uses bubblewrap too.

### General notes:

bubblewrap worked for me with the same instructions inside a Non-Qubes Debian jessie VirtualBox VM.

---

#### Related issues:

- https://trac.torproject.org/projects/tor/ticket/21077
- https://github.com/projectatomic/bubblewrap/issues/134

#### Workaround:

    sudo umount /proc/xen

**Comment by marmarek** [comment]
Do you have grsec kernel and/or apparmor enabled?

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by adrelanos** [comment]
Apparmor, yes.

**Comment by marmarek** [comment]
On Sun, Dec 25, 2016 at 01:31:31PM -0800, Patrick Schleizer wrote:
> Apparmor, yes.

Are you sure it isn't responsible for this denial? I have very little
experience with apparmor...

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by adrelanos** [comment]
Just now finished testing with apparmor removed kernel command line. (Verified using `sudo aa-status`.) Didn't fix this. 

-----

* grsecurity not installed.
* https://trac.torproject.org/projects/tor/wiki/doc/TorBrowser/Sandbox/Linux#Tickets
* Broken things that are unlikely to ever be supported:  - "Using the "hardened" series with a grsec kernel (ASAN/PaX conflict, not limited to the sandbox). "
* `Official 'sandboxed-tor-browser' binaries break with PaX MPROTECT.` - https://trac.torproject.org/projects/tor/ticket/20976

**Comment by adrelanos** [comment]
Also tested this also in a Qubes Debian stretch VM (upgrades a jessie template to stretch). Didn't fix this issue.

Also thought perhaps the kernel provided by dom0 might be incompatible with bubblewrap. To rule that out I switched that Qubes Debian stretch VM to Debian stretch linux-image-amd64 using [Qubes VM kernel](https://www.qubes-os.org/doc/managing-vm-kernel/) / pvgrub.

Also asked the bubblewrap maintainers for any insights:
https://github.com/projectatomic/bubblewrap/issues/134

**Comment by adrelanos** [comment]
Quote @marmarek https://github.com/projectatomic/bubblewrap/issues/134#issuecomment-271998694:
> Indeed after unmounting /proc/xen it does work. I wonder if anything still use /proc/xen in Qubes... AFAIR it's legacy location and the new one is /dev/xen. There were more problems with /proc/xen (where "normal files" behaves like character devices...). The fact that I could unmount it without killing anything suggests it isn't used anymore :)

That is great news!

/dev/xen is hardcoded a few times in multiple Qubes packages. Would solving this ticket now mean to change any mention of /dev/xen to /proc/xen? If so, should I create the pull requests to change this?

**Comment by marmarek** [comment]
> Would solving this ticket now mean to change any mention of /dev/xen to /proc/xen?

The other way around.
The great thing about `/dev/xen` is that it don't require to be specifically mounted (but still you require some modules being loaded) - so one potential race condition less.

**Comment by adrelanos** [comment]
Here is the first PR.

https://github.com/marmarek/qubes-builder-debian/pull/27

Does it look alright? If so, I would be going forward with the other components.

**Comment by adrelanos** [comment]
Actually, not as simple as a search, replace and pull request action as I thought. ( https://github.com/marmarek/qubes-builder-debian/pull/28 )

**Comment by marmarek** [comment]
Just tested - mostly works. The only problem I've found is missing `/proc/xen/capabilities`, which don't have corresponding file in `/dev/xen`. It's used [here](https://github.com/QubesOS/qubes-linux-utils/blob/master/udev/udev-block-add-change#L49) for checking if running in dom0 or VM. And probably few more places for the same purpose.

I'm looking at `/sys/hypervisor/properties/features`, where 0x800 bit is missing in VM (compared to the same in dom0). In theory it's [`XENFEAT_dom0`](https://github.com/xen-project/xen/blob/staging/xen/include/public/features.h#L95), but it's described as "operation as Dom0 is supported" - so, not necessary _running_ as dom0 (at least this is my understanding). Looking at the code setting this flag, it's totally not obvious if this is the same.

**Comment by adrelanos** [comment]
Maybe some dom0 package should ship (or already has a suitable) status
file that only exists in dom0? And then use that file to detect if
running inside dom0 or VM?

**Comment by alex-mazzariol** [comment]
This issue affects a lot of flatpak-distributed software (i.e. MonoDevelop), not only sandboxed TOR browser. It seems that flatpak internally uses bubblewrap too.

**Comment by ideologysec** [comment]
seems like this is [fixed upstream as of kernel 4.10.](https://github.com/containers/bubblewrap/issues/134#issuecomment-339049681)

can this be closed?

**Comment by alex-mazzariol** [comment]
Flatpak software is working as of now on fedora 32 templates with 4.19 kernel; the issue can be closed for me.
